plugins/processors/gpuattributes/processor.go

```diff
@@ -0,0 +1,195 @@
+//   - ClusterName
+//   - ClusterName, Namespace
+//   - ClusterName, Namespace, Service
+//   - ClusterName, Namespace, PodName
+//   - ClusterName, Namespace, PodName, FullPodName
+//   - ClusterName, Namespace, PodName, FullPodName, GpuDevice
+//
+// - Node
+//   - ClusterName
+//   - ClusterName, InstanceIdKey, NodeName
+//   - ClusterName, InstanceIdKey, NodeName, GpuDevice
+var nodeLabels = []string{
+	containerinsightscommon.ClusterNameKey,
+	containerinsightscommon.InstanceIdKey,
+	containerinsightscommon.GpuDeviceKey,
+	containerinsightscommon.MetricType,
+	containerinsightscommon.NodeNameKey,
+	containerinsightscommon.VersionKey,
+	containerinsightscommon.SourcesKey,
+	containerinsightscommon.Timestamp,
+}
+var podLabels = append([]string{
+	containerinsightscommon.K8sNamespace,
+	containerinsightscommon.FullPodNameKey,
+	containerinsightscommon.PodNameKey,
+	containerinsightscommon.TypeService,
+	containerinsightscommon.GpuUniqueId,
+}, nodeLabels...)
+var containerLabels = append([]string{
+	containerinsightscommon.ContainerNamekey,
+}, podLabels...)
+
+var nodeK8sLabels = []string{containerinsightscommon.HostKey}
+var podK8sLabels = append([]string{
+	"host",
+	"labels",
+	"pod_id",
+	"pod_name",
+	"pod_owners",
+	"namespace",
+}, nodeK8sLabels...)
+var containerK8sLabels = append([]string{
+	"container_name",
+	"containerd",
+}, podK8sLabels...)
+
+type gpuAttributesProcessor struct {
+	*Config
+	logger *zap.Logger
+}
+
+func newGpuAttributesProcessor(config *Config, logger *zap.Logger) *gpuAttributesProcessor {
+	d := &gpuAttributesProcessor{
+		Config: config,
+		logger: logger,
+	}
+	return d
+}
+
+func (d *gpuAttributesProcessor) processMetrics(_ context.Context, md pmetric.Metrics) (pmetric.Metrics, error) {
+	rms := md.ResourceMetrics()
+	for i := 0; i < rms.Len(); i++ {
+		rs := rms.At(i)
+		ilms := rs.ScopeMetrics()
+		for j := 0; j < ilms.Len(); j++ {
+			ils := ilms.At(j)
+			metrics := ils.Metrics()
+			for k := 0; k < metrics.Len(); k++ {
+				m := metrics.At(k)
+				d.processMetricAttributes(m)
+			}
+		}
+	}
+	return md, nil
+}
+
+func (d *gpuAttributesProcessor) processMetricAttributes(m pmetric.Metric) {
+	// only decorate GPU metrics
+	if !strings.Contains(m.Name(), gpuMetricIdentifier) {
+		return
+	}
+
+	var labels, k8sBlobLabels []string
+	if strings.HasPrefix(m.Name(), gpuContainerMetricPrefix) {
+		labels = containerLabels
+		k8sBlobLabels = containerK8sLabels
+	} else if strings.HasPrefix(m.Name(), gpuPodMetricPrefix) {
+		labels = podLabels
+		k8sBlobLabels = podK8sLabels
+	} else if strings.HasPrefix(m.Name(), gpuNodeMetricPrefix) {
+		labels = nodeLabels
+		k8sBlobLabels = nodeK8sLabels
+	}
+
+	labelFilter := map[string]map[string]interface{}{}
+	for _, attr := range labels {
+		labelFilter[attr] = nil
+	}
+	k8sBlobMap := map[string]interface{}{}
+	for _, attr := range k8sBlobLabels {
+		k8sBlobMap[attr] = nil
+	}
+	if len(k8sBlobMap) > 0 {
+		labelFilter[containerinsightscommon.K8sKey] = k8sBlobMap
+	}
```

Comment:
The whole `labelFilter` can be pre-computed ahead of time, not just the components.

Reply:
Very true. I was trying to avoid long vars that look very similar, but it's not worth doing adhoc generations of same data structure for every metrics.